parseComCode/main.js

```diff
@@ -30,6 +30,61 @@ Parse.Cloud.job("populateSearch", function(request, status) {
   });
 });
 
+// This job will remove any language records which currently do not have any books which use them.
+// The purpose is to keep BloomLibrary.org from displaying languages with no books.
+// It is scheduled to run every day.
+// It can be run manually using the following from the command line:
+// curl -X POST -H "X-Parse-Application-Id: <insert app ID>"  -H "X-Parse-Master-Key: <insert master key>" -d '{}' https://api.parse.com/1/jobs/removeUnusedLanguages
+// (In theory, the -d '{}' shouldn't be needed because we are not passing any parameters, but it doesn't work without it.)
+Parse.Cloud.job("removeUnusedLanguages", function(request, status) {
+    // Set up to modify data
+    Parse.Cloud.useMasterKey();
+
+    var allLangQuery = new Parse.Query('language');
+    allLangQuery.find({
+        success: function(languages) {
+            for (var i = 0; i < languages.length; i++) {
+                var language = languages[i];
+
+                var bookQuery = new Parse.Query('books');
+                bookQuery.equalTo('langPointers', language);
+
+                var languagesProcessed = 0;
+
+                // Use IIFE to pass the correct language to the callback
+                (function() {
+                    var lang = language;
+                    var numLanguages = languages.length;
+                    bookQuery.count({
+                        success: function (count) {
+                            if (count == 0) {
+                                lang.destroy({
+                                    success: function(deletedLanguage) {
+                                        console.log("Language " + deletedLanguage.get('name') + " was deleted because no books use it.");
+                                        if (++languagesProcessed === numLanguages)
+                                            status.success("removeUnusedLanguages completed successfully.");
+                                    },
+                                    error: function (deletedLanguage, error) {
+                                        status.error("Error in removeUnusedLanguages while deleting language " + deletedLanguage.get('name') + ": " + error);
+                                    }
+                                });
+                            } else if (++languagesProcessed === numLanguages) {
```

Comment:
Are these queries proceeding in parallel? If so, don't we need to worry about mutex on this variable?